<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  <title>Lost JavaScript</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  <meta content="website" property="og:type">
<meta property="og:title">
<meta content="http://lostjs.com/page/2/" property="og:url">
<meta property="og:image">
<meta content="Lost JavaScript" property="og:site_name">
<meta property="og:description">
<meta content="summary" name="twitter:card">
  
    <link rel="alternative" href="/atom.xml" title="Lost JavaScript" type="application/atom+xml">
  
  
    <link rel="icon" href="/favicon.png">
  
  <link href="http://fonts.googleapis.com/css?family=Source+Code+Pro" rel="stylesheet" type="text/css">
  <link rel="stylesheet" href="/css/style.css" type="text/css">
  <!--[if lt IE 9]><script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script><![endif]-->
  
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-450827-9']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


</head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">Lost JavaScript</a>
      </h1>
      
        <h2 id="subtitle-wrap">
          <a href="/" id="subtitle">一个 Pure JavaScript 工程师的博客</a>
        </h2>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
      </nav>
      <nav id="sub-nav">
        
          <a id="nav-rss-link" class="nav-icon" href="/atom.xml" title="RSS Feed"></a>
        
        <a id="nav-search-btn" class="nav-icon" title="搜索"></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" results="0" class="search-form-input" placeholder="Search"><input type="submit" value="&#xF002;" class="search-form-submit"><input type="hidden" name="q" value="site:http://lostjs.com"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-a-simple-promise-cache-by-lru-cache" class="article article-type-post" itemscope itemprop="blogPost">
  <div class="article-meta">
    <a href="/2014/04/21/a-simple-promise-cache-by-lru-cache/" class="article-date">
  <time datetime="2014-04-21T14:01:00.000Z" itemprop="datePublished">2014年04月21日</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="article-title" href="/2014/04/21/a-simple-promise-cache-by-lru-cache/">异步、promise 与缓存</a>
    </h1>
  

      </header>
    
    <div class="article-entry" itemprop="articleBody">
      
        <h2 id="异步操作缓存">异步操作缓存</h2>
<p>在 Node 开发过程中要常常和异步调用打交道，例如发送 http 请求、读取文件内容等。有时我们需要将这些异步 I/O 操作的结果缓存下来，使得程序运行的速度更快，例如:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
</pre></td><td class="code"><pre><span class="keyword">var</span> fileContent = <span class="literal">null</span>;
<span class="function"><span class="keyword">function</span> <span class="title">readFile</span><span class="params">(callback)</span> {</span>
    <span class="keyword">if</span> (fileContent !== <span class="literal">null</span>) {
        process.nextTick(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
            callback(<span class="literal">null</span>, fileContent);
        });
        <span class="keyword">return</span>;
    }

    fs.readFile(FILE_PATH, <span class="function"><span class="keyword">function</span> <span class="params">(err, content)</span> {</span>
        <span class="keyword">if</span> (err) {
            callback(err);
            <span class="keyword">return</span>;
        }

        fileContent = content;
        callback(<span class="literal">null</span>, content);
    });
}
</pre></td></tr></table></figure>

<p>上面的代码用 <code>fileContent</code> 变量做了一个简单的内存缓存，在 <code>readFile</code> 函数中，如果发现缓存中存在内容，则跳过文件读取操作。</p>
<p>这是个非常简单的缓存应用案例，我们可以将代码中的缓存逻辑抽出来，与业务逻辑分离，成为一个通用的缓存方法 <code>cacheAsync</code>:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
</pre></td><td class="code"><pre><span class="function"><span class="keyword">function</span> <span class="title">cacheAsync</span><span class="params">(logical)</span> {</span>
    <span class="keyword">var</span> cache = <span class="literal">null</span>;
    <span class="keyword">return</span> <span class="function"><span class="keyword">function</span> <span class="params">(callback)</span> {</span>
        <span class="keyword">if</span> (cache) {
            process.nextTick(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
                callback(<span class="literal">null</span>, cache);
            });
            <span class="keyword">return</span>;
        }

        logical(<span class="function"><span class="keyword">function</span> <span class="params">(err, result)</span> {</span>
            <span class="keyword">if</span> (err) {
                logical(err);
                <span class="keyword">return</span>;
            }

            cache = result;
            callback(<span class="literal">null</span>, result);
        });
    };
}

<span class="keyword">var</span> readFile = cacheAsync(<span class="function"><span class="keyword">function</span> <span class="params">(callback)</span> {</span>
    fs.readFile(FILE_PATH, callback);
});

readFile(<span class="function"><span class="keyword">function</span> <span class="params">(err, content)</span> {</span>
    <span class="keyword">if</span> (err) {
        <span class="keyword">throw</span> err;
        <span class="keyword">return</span>;
    }

    console.log(<span class="string">'file:'</span>, content);
});
</pre></td></tr></table></figure>

<p>使用 <code>cacheAsync</code> 方法可以很容易的给一个异步回调函数加上缓存。</p>
<h2 id="promise">promise</h2>
<p><code>promise</code> 是和上面代码中示范的 callback 异步回调方式截然不同的另一种异步范式。<code>promise</code> 范式下的异步函数不再接收一个 callback 函数，而是返回一个 promise 对象。promise 对象通过 <code>then</code> 方法来绑定回调函数，通过 <code>catch</code> 方法绑定错误处理函数。</p>
<p>nodejs 目前有很多个 promise 异步范式的库，最流行的是 <code>q</code>。这里以 q 为例，示范一个简单的 promise 异步范式:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
</pre></td><td class="code"><pre><span class="keyword">var</span> readFile = <span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
    <span class="comment">// q.nfcall 使一个 callback 风格的异步函数返回一个 promise 对象</span>
    <span class="keyword">return</span> q.nfcall(fs.readFile, FILE_PATH);
};

<span class="comment">// 下面这行看起来就和同步调用一样</span>
<span class="keyword">var</span> fileContent = readFile();

fileContent.then(<span class="function"><span class="keyword">function</span> <span class="params">(content)</span> {</span>
    console.log(<span class="string">'file:'</span>, content);
}).catch(<span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span>
    console.log(err);
});
</pre></td></tr></table></figure>

<p>看起来是不是感觉代码更复杂了？除了能看到异常逻辑和正常流程被分离了以外，似乎没有太多好处？那么看一个更复杂的例子，这个例子尝试从两个数据源读取用户信息:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
</pre></td><td class="code"><pre><span class="function"><span class="keyword">function</span> <span class="title">getUserSync</span><span class="params">()</span> {</span>
    <span class="keyword">var</span> user = fetchUserFromWeiboSync();

    <span class="keyword">if</span> (!user) {
        user = fetchUserFromRenrenSync();
    }

    <span class="keyword">if</span> (!user) {
        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="built_in">Error</span>(<span class="number">404</span>);
    }

    <span class="keyword">return</span> user;
}
</pre></td></tr></table></figure>

<p>callback 回调版本，一个常见的 callback hell:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
</pre></td><td class="code"><pre><span class="function"><span class="keyword">function</span> <span class="title">getUserAsync</span><span class="params">(callback)</span> {</span>
    fetchUserFromWeiboAsync(<span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span>
        <span class="keyword">if</span> (err) {
            fetchUserFromRenrenAsync(<span class="function"><span class="keyword">function</span> <span class="params">(err, user)</span> {</span>
                <span class="keyword">if</span> (err) {
                    callback(<span class="keyword">new</span> <span class="built_in">Error</span>(<span class="number">404</span>));
                } <span class="keyword">else</span> {
                    callback(<span class="literal">null</span>, user);
                }
            });
        } <span class="keyword">else</span> {
            callback(<span class="literal">null</span>, user);
        }
    })
}
</pre></td></tr></table></figure>

<p>promise 版本，代码之少令人惊讶:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
</pre></td><td class="code"><pre><span class="function"><span class="keyword">function</span> <span class="title">getUserQ</span><span class="params">()</span> {</span>
    <span class="keyword">return</span> fetchUserFromWeiboQ().catch(fetchUserFromRenrenQ);
}
</pre></td></tr></table></figure>

<h2 id="promise_与缓存">promise 与缓存</h2>
<p>给一个 promise 范式下的异步函数加缓存是一件非常轻松的事情，因为你不需要处理任何的异常流程:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
</pre></td><td class="code"><pre><span class="keyword">var</span> q = <span class="built_in">require</span>(<span class="string">'q'</span>);

<span class="function"><span class="keyword">function</span> <span class="title">cacheQ</span><span class="params">(logical)</span> {</span>
    <span class="keyword">var</span> cache = <span class="literal">null</span>;

    <span class="keyword">return</span> <span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
        <span class="keyword">if</span> (cache !== <span class="literal">null</span>) {
            <span class="keyword">return</span> q(cache);
        }

        <span class="comment">// 这里不需要再处理任何的异常，缓存方法和程序逻辑进一步解耦</span>
        <span class="keyword">return</span> logical().then(<span class="function"><span class="keyword">function</span> <span class="params">(result)</span> {</span>
            cache = result;
            <span class="comment">// 这里必须 return，因为这里的返回值会作为参数传递给下一个 then 方法绑定的回调函数</span>
            <span class="keyword">return</span> result;
        });
    };
}

<span class="keyword">var</span> readFile = cacheQ(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
    <span class="comment">// q.nfcall 使一个 callback 风格的异步函数返回一个 promise 对象</span>
    <span class="keyword">return</span> q.nfcall(fs.readFile, FILE_PATH);
});
</pre></td></tr></table></figure>

<h2 id="更灵活的缓存">更灵活的缓存</h2>
<p>在实际应用中，一个缓存函数还应当支持以下特性:</p>
<ol>
<li>缓存应该是一个 key-value 存储，而不是只能缓存一个值</li>
<li>避免使用一个简单的内存对象来做缓存，更好的选择有 lru-cache、redis 以及 memcached 等</li>
<li>缓存函数应该对逻辑函数透明，即逻辑函数是否被缓存，不应该影响整个程序的执行结果</li>
</ol>
<p>以下给一个实际项目中应用的例子:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
</pre></td><td class="code"><pre><span class="keyword">var</span> q = <span class="built_in">require</span>(<span class="string">'q'</span>);
<span class="keyword">var</span> LRU = <span class="built_in">require</span>(<span class="string">'lru-cache'</span>);

<span class="function"><span class="keyword">function</span> <span class="title">cacheQ</span><span class="params">(logical, key)</span> {</span>
    <span class="keyword">var</span> cache = <span class="keyword">new</span> LRU({
        max: <span class="number">100</span>, <span class="comment">// 最大缓存数量，防止内存泄露</span>
        maxAge: <span class="number">60000</span> <span class="comment">// 一分钟过期</span>
    });

    <span class="keyword">return</span> <span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
        <span class="keyword">var</span> args = <span class="built_in">arguments</span>;
        key = <span class="keyword">typeof</span> key === <span class="string">'function'</span> ? key.apply(<span class="keyword">this</span>, args) : <span class="string">'__default__'</span>;

        <span class="comment">// 这里支持 key 作为一个 promise 对象，所以需要用 q 来包装这个对象</span>
        <span class="keyword">return</span> q(key).then(<span class="function"><span class="keyword">function</span> <span class="params">(key)</span> {</span>
            <span class="keyword">if</span> (cache.has(key)) {
                <span class="keyword">return</span> cache.get(key);
            }

            <span class="keyword">return</span> logical.apply(<span class="keyword">this</span>, args).then(<span class="function"><span class="keyword">function</span> <span class="params">(result)</span> {</span>
                cache.set(key, result);
                <span class="keyword">return</span> result;
            });
        }.bind(<span class="keyword">this</span>)); <span class="comment">// 注意 this 的传递</span>
    };
}

<span class="keyword">var</span> readFileCached = cacheQ(<span class="function"><span class="keyword">function</span> <span class="params">(path)</span> {</span>
    <span class="keyword">return</span> q.nfcall(fs.readFile, path);
}, <span class="function"><span class="keyword">function</span> <span class="params">(path)</span> {</span>
    <span class="keyword">return</span> path;
});
</pre></td></tr></table></figure>


        <p>- FIN -</p>
      
    </div>
    
    <footer class="article-footer">
      <a data-url="http://lostjs.com/2014/04/21/a-simple-promise-cache-by-lru-cache/" data-id="vrf7u9mwtslxxtny" class="article-share-link">分享</a>
      
      
        <a href="http://lostjs.com/2014/04/21/a-simple-promise-cache-by-lru-cache/#comments" class="article-comment-link">评论</a>
      
      
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/node/">node</a></li></ul>

    </footer>
  </div>
  
</article>




  
    <article id="post-docker-behind-gfw" class="article article-type-post" itemscope itemprop="blogPost">
  <div class="article-meta">
    <a href="/2014/02/26/docker-behind-gfw/" class="article-date">
  <time datetime="2014-02-26T03:16:00.000Z" itemprop="datePublished">2014年02月26日</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="article-title" href="/2014/02/26/docker-behind-gfw/">Mac 下使用代理连接 docker 镜像服务器</a>
    </h1>
  

      </header>
    
    <div class="article-entry" itemprop="articleBody">
      
        <p>Docker 是个非常牛逼的轻量级虚拟化解决方案，它可以将整个操作系统打成一个非常小的镜像，并且直接跑在宿主的 Linux Kernel 之上。然而在国内因为可恶的 GFW 存在，使用 docker 成了一件非常困难的事情。具体表现在使用 <code>docker search</code> 和 <code>docker pull</code> 的时候连接非常不稳定，这一刻我又深深的感觉到了作为一名天朝程序员的悲哀…</p>
<h3 id="使用_HTTP_代理连接_docker_中央服务器">使用 HTTP 代理连接 docker 中央服务器</h3>
<p>如果你的宿主操作系统是 linux 那方法就很简单了，直接通过 <code>HTTP_PROXY=127.0.0.1:8118 HTTPS_PROXY=127.0.0.1:8118 docker -d</code> 的形式启动服务即可.</p>
<h3 id="boot2docker_使用代理的方法"><code>boot2docker</code> 使用代理的方法</h3>
<p>Mac 下使用 docker 就必须用 <code>boot2docker</code> 来创建一个 Virtual Box 虚拟机提供 Linux 宿主环境，要用代理就略微折腾一点。如果只是临时的想用一下，可以通过以下命令来搞定:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre>1
2
3
4
5
</pre></td><td class="code"><pre>boot2docker ssh <span class="comment"># 密码是 tcuser</span>
<span class="built_in">sudo</span> vi /var/lib/boot2docker/profile
<span class="comment"># export HTTP_PROXY=10.0.1.64:8118</span>
<span class="comment"># export HTTPS_PROXY=10.0.1.64:8118</span>
/usr/local/etc/init.d/docker restart
</pre></td></tr></table></figure>


        <p>- FIN -</p>
      
    </div>
    
    <footer class="article-footer">
      <a data-url="http://lostjs.com/2014/02/26/docker-behind-gfw/" data-id="6pa6a9c160gxkl2k" class="article-share-link">分享</a>
      
      
        <a href="http://lostjs.com/2014/02/26/docker-behind-gfw/#comments" class="article-comment-link">评论</a>
      
      
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/docker/">docker</a></li></ul>

    </footer>
  </div>
  
</article>




  
    <article id="post-handle-exception-in-node" class="article article-type-post" itemscope itemprop="blogPost">
  <div class="article-meta">
    <a href="/2014/01/25/handle-exception-in-node/" class="article-date">
  <time datetime="2014-01-25T04:18:00.000Z" itemprop="datePublished">2014年01月25日</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="article-title" href="/2014/01/25/handle-exception-in-node/">Node 出现 uncaughtException 之后的优雅退出方案</a>
    </h1>
  

      </header>
    
    <div class="article-entry" itemprop="articleBody">
      
        <p>Node 的异步特性是它最大的魅力，但是在带来便利的同时也带来了不少麻烦和坑，错误捕获就是一个。由于 Node 的异步特性，导致我们无法使用 try/catch 来捕获回调函数中的异常，例如:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
</pre></td><td class="code"><pre><span class="keyword">try</span> {
    console.log(<span class="string">'进入 try/catch'</span>);
    <span class="built_in">require</span>(<span class="string">'fs'</span>).stat(<span class="string">'SOME_FILE_DOES_NOT_EXIST'</span>, <span class="function"><span class="keyword">function</span> <span class="title">readCallback</span><span class="params">(err, content)</span> {</span>
        <span class="keyword">if</span> (err) {
            <span class="keyword">throw</span> err; <span class="comment">// 抛出异常</span>
        }
    });
} <span class="keyword">catch</span> (e) {
    <span class="comment">// 这里捕获不到 readCallback 函数中抛出的异常</span>
} <span class="keyword">finally</span> {
    console.log(<span class="string">'离开 try/catch'</span>);
}
</pre></td></tr></table></figure>

<p>运行结果是:</p>
<figure class="highlight"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
</pre></td><td class="code"><pre>进入 <span class="keyword">try</span>/<span class="keyword">catch</span>
离开 <span class="keyword">try</span>/<span class="keyword">catch</span>

test.js:<span class="number">7</span>
            <span class="keyword">throw</span> err; <span class="comment">// 抛出异常</span>
                  ^
<span class="built_in">Error</span>: ENOENT, stat <span class="string">'SOME_FILE_DOES_NOT_EXIST'</span>
</pre></td></tr></table></figure>

<p>上面代码中由于 <code>fs.stat</code> 去查询一个不存在的文件的状态，导致 <code>readCallback</code> 抛出了一个异常。由于 <code>fs.read</code> 的异步特性，<code>readCallback</code> 函数的调用发生在 <code>try/catch</code> 块结束之后，所以该异常不会被 try/catch 捕获。之后 Node 会触发 <code>uncaughtException</code> 事件，如果这个事件依然没有得到响应，整个进程(<code>process</code>)就会 crash。</p>
<p>程序员永远无法保证代码中不出现 <code>uncaughtException</code>，即便是自己代码写的足够小心，也不能保证用的第三方模块没有 bug，例如:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
</pre></td><td class="code"><pre><span class="keyword">var</span> deserialize = <span class="built_in">require</span>(<span class="string">'deserialize'</span>); <span class="comment">// 假设 deserialize 是一个带有 bug 的第三方模块</span>

<span class="comment">// app 是一个 express 服务对象</span>
app.get(<span class="string">'/users'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(req, res)</span> {</span>
    mysql.query(<span class="string">'SELECT * FROM user WHERE id=1'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err, user)</span> {</span>
        <span class="keyword">var</span> config = deserialize(user.config); <span class="comment">// 假如这里触发了 deserialize 的 bug</span>
        res.send(config);
    });
});
</pre></td></tr></table></figure>

<p>如果不幸触发了 <code>deserialize</code> 模块的 bug，这里就会抛出一个异常，最终结果是整个服务 crash。</p>
<p>当这种情况发生在 Web 服务上时结果是灾难性的。<code>uncaughtException</code> 错误会导致当前的所有的用户连接都被中断，甚至不能返回一个正常的 HTTP 错误码，用户只能等到浏览器超时才能看到一个 <code>no data received</code> 错误。</p>
<p>这是一种非常野蛮粗暴的异常处理机制，任何线上服务都不应该因为 <code>uncaughtException</code> 导致服务器崩溃。一个友好的错误处理机制应该满足三个条件:</p>
<ol>
<li>对于引发异常的用户，返回 500 页面</li>
<li>其他用户不受影响，可以正常访问</li>
<li>不影响整个进程的正常运行</li>
</ol>
<p>很遗憾的是，保证 <code>uncaughtException</code> 不影响整个进程的健康运转是不可能的。当 Node 抛出 <code>uncaughtException</code> 异常时就会丢失当前环境的堆栈，导致 Node 不能正常进行内存回收。也就是说，每一次 <code>uncaughtException</code> 都有可能导致内存泄露。</p>
<p>既然如此，退而求其次，我们可以在满足前两个条件的情况下退出进程以便重启服务。</p>
<h2 id="用_domain_来捕获异步异常">用 domain 来捕获异步异常</h2>
<p>普遍的思路是，如果可以通过某种方式来捕获回调函数中的异常，那么就不会有 <code>uncaughtException</code> 错误导致的崩溃。为了解决这个问题，Node 0.8 之后的版本新增了 <code>domain</code> 模块，它可以用来捕获回调函数中抛出的异常。</p>
<p><code>domain</code> 主要的 API 有 <code>domain.run</code> 和 <code>error</code> 事件。简单的说，通过 <code>domain.run</code> 执行的函数中引发的异常都可以通过 <code>domain</code> 的 <code>error</code> 事件捕获，例如:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
</pre></td><td class="code"><pre><span class="keyword">var</span> domain = <span class="built_in">require</span>(<span class="string">'domain'</span>);
<span class="keyword">var</span> d = domain.create();
d.run(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
    setTimeout(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="built_in">Error</span>(<span class="string">'async error'</span>); <span class="comment">// 抛出一个异步异常</span>
    }, <span class="number">1000</span>);
});

d.on(<span class="string">'error'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span>
    console.log(<span class="string">'catch err:'</span>, err); <span class="comment">// 这里可以捕获异步异常</span>
});
</pre></td></tr></table></figure>

<p>通过 <code>domain</code> 模块，以及 JavaScript 的词法作用域特性，可以很轻易的为引发异常的用户返回 500 页面。以 express 为例:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
</pre></td><td class="code"><pre><span class="keyword">var</span> app = express();
<span class="keyword">var</span> server = <span class="built_in">require</span>(<span class="string">'http'</span>).createServer(app);
<span class="keyword">var</span> domain = <span class="built_in">require</span>(<span class="string">'domain'</span>);

app.use(<span class="function"><span class="keyword">function</span> <span class="params">(req, res, next)</span> {</span>
    <span class="keyword">var</span> reqDomain = domain.create();
    reqDomain.on(<span class="string">'error'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span> <span class="comment">// 下面抛出的异常在这里被捕获</span>
        res.send(<span class="number">500</span>, err.stack); <span class="comment">// 成功给用户返回了 500</span>
    });

    reqDomain.run(next);
});

app.get(<span class="string">'/'</span>, <span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
    setTimeout(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="built_in">Error</span>(<span class="string">'async exception'</span>); <span class="comment">// 抛出一个异步异常</span>
    }, <span class="number">1000</span>);
});
</pre></td></tr></table></figure>

<p>上面的代码将 domain 作为一个中间件来使用，保证之后 express 所有的中间件都在 <code>domain.run</code> 函数内部执行。这些中间件内的异常都可以通过 <code>error</code> 事件来捕获。</p>
<p>尽管借助于闭包，我们可以正常的给用户返回 500 错误，但是 <code>domain</code> 捕获到错误时依然会丢失堆栈信息，此时已经无法保证程序的健康运行，必须退出。Node http server 提供了 <code>close</code> 方法，该方法在调用时会停止 server 接收新的请求，但不会断开当前已经建立的连接。</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
</pre></td><td class="code"><pre>reqDomain.on(<span class="string">'error'</span>, <span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
    <span class="keyword">try</span> {
        <span class="comment">// 强制退出机制</span>
        <span class="keyword">var</span> killTimer = setTimeout(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
            process.exit(<span class="number">1</span>);
        }, <span class="number">30000</span>);
        killTimer.unref(); <span class="comment">// 非常重要</span>

        <span class="comment">// 自动退出机制，停止接收新链接，等待当前已建立连接的关闭</span>
        server.close(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
            <span class="comment">// 此时所有连接均已关闭，此时 Node 会自动退出，不需要再调用 process.exit(1) 来结束进程</span>
        });
    } <span class="keyword">catch</span>(e) {
        console.log(<span class="string">'err'</span>, e.stack);
    }
});
</pre></td></tr></table></figure>

<p>这个例子来自 Node 的文档。其中有几个关键点：</p>
<ul>
<li>Node 有个非常好的特性，所有连接都被释放后进程会自动结束，所以不需要再 <code>server.close</code> 方法的回调函数中退出进程</li>
<li>强制退出机制: 因为用户连接有可能因为某些原因无法释放，在这种情况下应该强制退出整个进程。</li>
<li><code>killTimer.unref()</code>: 如果不使用 <code>unref</code> 方法，那么即使 server 的所有连接都关闭，Node 也会保持运行直到 <code>killTimer</code> 的回调函数被调用。<code>unref</code> 可以创建一个”不保持程序运行”的计时器。</li>
<li>处理异常时要小心的把异常处理逻辑用 try/catch 包住，避免处理异常时抛出新的异常</li>
</ul>
<p>通过 <code>domain</code> 似乎就已经解决了我们的需求: 给触发异常的用户一个 500，停止接收新请求，提供正常的服务给已经建立连接的用户，直到所有请求都已结束，退出进程。但是，理想很丰满，现实很骨感，<code>domain</code> 有个最大的问题，它<a href="http://cnodejs.org/topic/516b64596d38277306407936" target="_blank">不能捕获所有的异步异常</a>！。也就是说，即使用了 <code>domain</code>，程序依然有因为 <code>uncaughtException</code> crash 的可能。</p>
<p>所幸的是我们可以监听 <code>uncaughtException</code> 事件。</p>
<h2 id="uncaughtException_事件"><code>uncaughtException</code> 事件</h2>
<p><code>uncaughtException</code> 是一个非常古老的事件。当 Node 发现一个未捕获的异常时，会触发这个事件。并且如果这个事件存在回调函数，Node 就不会强制结束进程。这个特性，可以用来弥补 <code>domain</code> 的不足:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
</pre></td><td class="code"><pre>process.on(<span class="string">'uncaughtException'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span>
    console.log(err);

    <span class="keyword">try</span> {
        <span class="keyword">var</span> killTimer = setTimeout(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
            process.exit(<span class="number">1</span>);
        }, <span class="number">30000</span>);
        killTimer.unref();

        server.close();
    } <span class="keyword">catch</span> (e) {
        console.log(<span class="string">'error when exit'</span>, e.stack);
    }
});
</pre></td></tr></table></figure>

<p><code>uncaughtException</code> 事件的缺点在于无法为抛出异常的用户请求返回一个 500 错误，这是由于 <code>uncaughtException</code> 丢失了当前环境的上下文，比如下面的例子就是它做不到的:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
</pre></td><td class="code"><pre>app.get(<span class="string">'/'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(req, res)</span> {</span>
    setTimeout(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="built_in">Error</span>(<span class="string">'async error'</span>); <span class="comment">// uncaughtException, 导致 req 的引用丢失</span>
        res.send(<span class="number">200</span>);
    }, <span class="number">1000</span>);
});

process.on(<span class="string">'uncaughtException'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span>
    res.send(<span class="number">500</span>); <span class="comment">// 做不到，拿不到当前请求的 res 对象</span>
});
</pre></td></tr></table></figure>

<p>最终出错的用户只能等待浏览器超时。</p>
<h2 id="domain_+_uncaughtException"><code>domain</code> + <code>uncaughtException</code></h2>
<p>所以，我们可以结合两种异常捕获机制，用 <code>domain</code> 来捕获大部分的异常，并且提供友好的 500 页面以及优雅退出。对于剩下的异常，通过 <code>uncaughtException</code> 事件来避免服务器直接 crash。</p>
<p>代码如下:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
</pre></td><td class="code"><pre><span class="keyword">var</span> app = express();
<span class="keyword">var</span> server = <span class="built_in">require</span>(<span class="string">'http'</span>).create(app);
<span class="keyword">var</span> domain = <span class="built_in">require</span>(<span class="string">'domain'</span>);

<span class="comment">// 使用 domain 来捕获大部分异常</span>
app.use(<span class="function"><span class="keyword">function</span> <span class="params">(req, res, next)</span> {</span>
    <span class="keyword">var</span> reqDomain = domain.create();
    reqDomain.on(<span class="string">'error'</span>, <span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
        <span class="keyword">try</span> {
            <span class="keyword">var</span> killTimer = setTimeout(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
                process.exit(<span class="number">1</span>);
            }, <span class="number">30000</span>);
            killTimer.unref();

            server.close();

            res.send(<span class="number">500</span>);
        } <span class="keyword">catch</span> (e) {
            console.log(<span class="string">'error when exit'</span>, e.stack);
        }
    });

    reqDomain.run(next);
});

<span class="comment">// uncaughtException 避免程序崩溃</span>
process.on(<span class="string">'uncaughtException'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span>
    console.log(err);

    <span class="keyword">try</span> {
        <span class="keyword">var</span> killTimer = setTimeout(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
            process.exit(<span class="number">1</span>);
        }, <span class="number">30000</span>);
        killTimer.unref();

        server.close();
    } <span class="keyword">catch</span> (e) {
        console.log(<span class="string">'error when exit'</span>, e.stack);
    }
});
</pre></td></tr></table></figure>

<h2 id="其他的一些问题">其他的一些问题</h2>
<h3 id="express_中异常的处理"><code>express</code> 中异常的处理</h3>
<p>使用 <code>express</code> 时记住一定不要在 controller 的异步回调中抛出异常，例如:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
</pre></td><td class="code"><pre>app.get(<span class="string">'/'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(req, res, next)</span> {</span> <span class="comment">// 总是接收 next 参数</span>
    mysql.query(<span class="string">'SELECT * FROM users'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err, results)</span> {</span>
        <span class="comment">// 不要这样做</span>
        <span class="keyword">if</span> (err) <span class="keyword">throw</span> err;

        <span class="comment">// 应该将 err 传递给 errorHandler 处理</span>
        <span class="keyword">if</span> (err) <span class="keyword">return</span> next(err);
    });
});

app.use(<span class="function"><span class="keyword">function</span> <span class="params">(err, req, res, next)</span> {</span> <span class="comment">// 带有四个参数的 middleware 专门用来处理异常</span>
    res.render(<span class="number">500</span>, err.stack);
});
</pre></td></tr></table></figure>

<h3 id="和_cluster_一起使用">和 cluster 一起使用</h3>
<p>cluster 是 node 自带的负载均衡模块，使用 cluster 模块可以方便的建立起一套 master/slave 服务。在使用 cluster 模块时，需要注意不仅需要调用 <code>server.close()</code> 来关闭连接，同时还需要调用 <code>cluster.worker.disconnect()</code> 通知 master 进程已停止服务:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
</pre></td><td class="code"><pre><span class="keyword">var</span> cluster = <span class="built_in">require</span>(<span class="string">'cluster'</span>);

process.on(<span class="string">'uncaughtException'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span>
    console.log(err);

    <span class="keyword">try</span> {
        <span class="keyword">var</span> killTimer = setTimeout(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
            process.exit(<span class="number">1</span>);
        }, <span class="number">30000</span>);
        killTimer.unref();

        server.close();

        <span class="keyword">if</span> (cluster.worker) {
            cluster.worker.disconnect();
        }
    } <span class="keyword">catch</span> (e) {
        console.log(<span class="string">'error when exit'</span>, e.stack);
    }
});
</pre></td></tr></table></figure>

<h3 id="不要通过_uncaughtException_来忽略错误">不要通过 <code>uncaughtException</code> 来忽略错误</h3>
<p>当 <code>uncaughtException</code> 事件有一个以上的 <code>listener</code> 时，会阻止 Node 结束进程。因此就有一个广泛流传的做法是监听 <code>process</code> 的 <code>uncaughtException</code> 事件来阻止进程退出，这种做法有内存泄露的风险，所以千万不要这么做:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
</pre></td><td class="code"><pre>process.on(<span class="string">'uncaughtException'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span> <span class="comment">// 不要这么做</span>
    console.log(err);
});
</pre></td></tr></table></figure>

<h3 id="pm2_对于_uncaughtException_的额外处理">pm2 对于 <code>uncaughtException</code> 的额外处理</h3>
<p>如果你在用 pm2 0.7.1 之前的版本，那么要当心。pm2 有一个 bug，如果进程抛出了 <code>uncaughtException</code>，无论代码中是否捕获了这个事件，进程都会被 pm2 杀死。0.7.2 之后的 pm2 解决了这个问题。</p>
<h3 id="要小心_worker-disconnect()">要小心 worker.disconnect()</h3>
<p>如果你在退出进程时希望可以发消息给监控服务器，并且还使用了 cluster，那么这个时候要特别小心，比如下面的代码:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
</pre></td><td class="code"><pre><span class="keyword">var</span> udpLog = dgram.createSocket(<span class="string">'udp4'</span>);
<span class="keyword">var</span> cluster = <span class="built_in">require</span>(<span class="string">'cluster'</span>);

process.on(<span class="string">'uncaughtException'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span>
    udpLog.send(<span class="string">'process '</span> + process.pid + <span class="string">' down'</span>, <span class="comment">/* ... 一些发送 udp 消息的参数 ...*/</span>);

    server.close();
    cluster.worker.disconnect();
});
</pre></td></tr></table></figure>

<p>这份代码就不能正常的将消息发送出去。因为 <code>udpLog.send</code> 是一个异步方法，真正发消息的操作发生在下一个事件循环中。而在真正的发送消息之前 <code>cluster.worker.disconnect()</code> 就已经执行了。<code>worker.disconnect()</code> 会在当前进程没有任何链接之后，杀掉整个进程，这种情况有可能发生在发送 log 数据之前，导致 log 数据发不出去。</p>
<p>一个解决方法是在 <code>udpLog.send</code> 方法发送完数据后再调用 <code>worker.disconnect</code>:</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre>1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
</pre></td><td class="code"><pre><span class="keyword">var</span> udpLog = dgram.createSocket(<span class="string">'udp4'</span>);
<span class="keyword">var</span> cluster = <span class="built_in">require</span>(<span class="string">'cluster'</span>);

process.on(<span class="string">'uncaughtException'</span>, <span class="function"><span class="keyword">function</span> <span class="params">(err)</span> {</span>
    udpLog.send(<span class="string">'process '</span> + process.pid + <span class="string">' down'</span>, <span class="comment">/* ... 一些发送 udp 消息的参数 ...*/</span>, <span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
        cluster.worker.disconnect();
    });

    server.close();

    <span class="comment">// 保证 worker.disconnect 不会拖太久..</span>
    setTimeout(<span class="function"><span class="keyword">function</span> <span class="params">()</span> {</span>
        cluster.worker.disconnect();
    }, <span class="number">100</span>).unref();
});
</pre></td></tr></table></figure>

<h2 id="小节">小节</h2>
<p>说了这么多，结论是，目前为止(Node 0.10.25)，依然没有一个完美的方案来解决任意异常的优雅退出问题。用 <code>domain</code> 来捕获大部分异常，并且通过 <code>uncaughtException</code> 避免程序 crash 是目前来说最理想的方案。回调异常的退出问题在遇到 cluster 以后会更加复杂，特别是对于连接关闭的处理要格外小心。</p>
<h2 id="参考文章">参考文章</h2>
<ul>
<li><a href="http://nodejs.org/api/domain.html#domain_warning_don_t_ignore_errors" target="_blank">Don’t ignore errors</a></li>
<li><a href="http://nodejs.org/api/domain.html" target="_blank">Node API: domain</a></li>
<li><a href="http://nodejs.org/api/process.html" target="_blank">Node API: process</a></li>
<li><a href="http://cnodejs.org/topic/516b64596d38277306407936" target="_blank">Node.js 异步异常的处理与domain模块解析</a></li>
<li><a href="http://www.alloyteam.com/2013/12/node-js-series-exception-caught/" target="_blank">Node.js异常捕获的一些实践</a></li>
</ul>

        <p>- FIN -</p>
      
    </div>
    
    <footer class="article-footer">
      <a data-url="http://lostjs.com/2014/01/25/handle-exception-in-node/" data-id="7wke9pjvydksyzfr" class="article-share-link">分享</a>
      
      
        <a href="http://lostjs.com/2014/01/25/handle-exception-in-node/#comments" class="article-comment-link">评论</a>
      
      
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/node/">node</a></li></ul>

    </footer>
  </div>
  
</article>




  
  
    <nav id="page-nav">
      <a class="extend prev" href="/">&laquo; 上一页</a><a class="page-number" href="/">1</a><span class="page-number current">2</span><a class="page-number" href="/page/3/">3</a><a class="page-number" href="/page/4/">4</a><a class="page-number" href="/page/5/">5</a><a class="extend next" href="/page/3/">下一页 &raquo;</a>
    </nav>
  

</section>
        
          <aside id="sidebar">
  
    

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list"><li class="tag-list-item"><a class="tag-list-link" href="/tags/docker/">docker</a><span class="tag-list-count">1</span></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/git/">git</a><span class="tag-list-count">1</span></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/grunt/">grunt</a><span class="tag-list-count">1</span></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/hexo/">hexo</a><span class="tag-list-count">1</span></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/javascript/">javascript</a><span class="tag-list-count">1</span></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/mustache/">mustache</a><span class="tag-list-count">1</span></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/node/">node</a><span class="tag-list-count">5</span></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/vim/">vim</a><span class="tag-list-count">1</span></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/zsh/">zsh</a><span class="tag-list-count">1</span></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/招聘/">招聘</a><span class="tag-list-count">1</span></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/04">四月 2014</a><span class="archive-list-count">4</span></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/02">二月 2014</a><span class="archive-list-count">1</span></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/01">一月 2014</a><span class="archive-list-count">1</span></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/10">十月 2013</a><span class="archive-list-count">1</span></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/07">七月 2013</a><span class="archive-list-count">1</span></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/12">十二月 2012</a><span class="archive-list-count">1</span></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/09">九月 2012</a><span class="archive-list-count">1</span></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/08">八月 2012</a><span class="archive-list-count">2</span></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/02">二月 2012</a><span class="archive-list-count">1</span></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/12">十二月 2011</a><span class="archive-list-count">1</span></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2014/04/24/upgrade-express-4/">Express 4.0 升级手记</a>
          </li>
        
          <li>
            <a href="/2014/04/23/replace-nginx-by-node/">用 node 代替 nginx 做请求分发</a>
          </li>
        
          <li>
            <a href="/2014/04/22/from-octopress-to-hexo/">把 blog 从 octopress 迁移到 hexo</a>
          </li>
        
          <li>
            <a href="/2014/04/21/a-simple-promise-cache-by-lru-cache/">异步、promise 与缓存</a>
          </li>
        
          <li>
            <a href="/2014/02/26/docker-behind-gfw/">Mac 下使用代理连接 docker 镜像服务器</a>
          </li>
        
      </ul>
    </div>
  </div>


  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      &copy; 2014 Ethan (pw) Zhang<br>
      Powered by <a href="http://zespia.tw/hexo/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>
    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
</nav>
    


<script type="text/javascript">
var duoshuoQuery = {
    short_name: "lostjs"
};
(function () {
    var ds = document.createElement('script');
    ds.type = 'text/javascript';
    ds.async = true;
    ds.src = (document.location.protocol == 'https:' ? 'https:' : 'http:') + '//static.duoshuo.com/embed.js';
    ds.charset = 'UTF-8';
    (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(ds);
})();
</script>


<script src="//ajax.googleapis.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>


<link rel="stylesheet" href="/fancybox/jquery.fancybox.css" media="screen" type="text/css">
<script src="/fancybox/jquery.fancybox.pack.js"></script>


<script type="text/javascript" src="/js/script.js"></script>

  </div>
</body>
</html>